src/pcp/iostat/pcp-iostat.1
'\"! tbl | mmdoc
'\"macro stdmacro
.\"
.\" Copyright (c) 2014-2015 Red Hat.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PCP-IOSTAT 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmiostat\f1,
\f3pcp-iostat\f1 \- performance metrics i/o statistics tool
.\" literals use .B or \f3
.\" arguments use .I or \f2
.SH SYNOPSIS
\f3pcp\ iostat\f1
[\f3\-A\f1 \f2align\f1 \f3--align=\f2TIME\f1]
[\f3\-a\f1 \f2archive\f1 \f3--archive=\f2FILE\f1]
[\f3\-h\f1 \f2host\f1 \f3--host=\f2HOST\f1]
[\f3\-O\f1 \f2offset\f1 \f3--origin=\f2TIME\f1]
[\f3\-S\f1 \f2starttime\f1 \f3--start=\f2TIME\f1]
[\f3\-s\f1 \f2samples\f1 \f3--samples=\f2N\f1]
[\f3\-T\f1 \f2endtime\f1 \f3--finish=\f2TIME\f1]
[\f3\-t\f1 \f2interval\f1 \f3--interval=\f2DELTA\f1]
[\f3\-P\f1 \f2precision\f1 \f3--precision=\f2N\f1]
[\f3\-u\f1 \f3--no-interpolate\f1]
[\f3\-Z\f1 \f2timezone\f1 \f3--timezone=\f2TZ\f1]
[\f3\-z\f1 \f3--hostzone\f1]
[\f3\-?\f1 \f3--help\f1]
[\f3\-x\f1 [dm][,t][,h]\f1]
.SH DESCRIPTION
.de EX
.in +0.5i
.ie t .ft CB
.el .ft B
.ie t .sp .5v
.el .sp
.ta \\w' 'u*8
.nf
..
.de EE
.fi
.ie t .sp .5v
.el .sp
.ft R
.in
..
.B pcp-iostat
reports I/O statistics for scsi devices (by default) or device-mapper devices (if the \f3-x dm\f1 option is specified).
By default
.B pcp-iostat
reports live data for the local host but can also report for a remote host (\f3-h\fP) or from a previously captured PCP archive (\f3-a\fP).
.PP
The
.BR \-S ,
.BR \-T ,
.BR \-O
and
.B \-A
options may be used to define a time window to restrict the
samples retrieved, set an initial origin within the time window,
or specify a ``natural'' alignment of the sample times; refer to
.BR PCPIntro (1)
for a complete description of these options.
.PP
The other options which control the source, timing and layout of the information
reported by
.B pcp-iostat
are as follows:
.TP 5
.B \-a
Performance metric values are retrieved from the Performance Co-Pilot (PCP)
archive log file identified by the base name
.IR archive .
See also
.BR \-u .
.TP
.B \-h
Current performance metric values are retrieved from the nominated
.I host
machine.
.TP
.B \-s
The argument
.I samples
defines the number of samples to be retrieved and reported.
If
.I samples
is 0 or
.B \-s
is not specified, 
.B pcp-iostat
will sample and report continuously (in real time mode) or until the end
of the PCP archive (in archive mode).
.TP
.B \-t
The default update \f2interval\f1 may be set to something other than the
default 1 second.
The
.I interval
argument follows the syntax described in
.BR PCPIntro (1),
and in the simplest form may be an unsigned integer (the implied
units in this case are seconds).
The \f3-t\fP option is particularly useful when replaying large archives (\f3-a\fP option) that span several hours or even days.
In this case specifying a large
.I interval
(e.g. 1h for 1 hour)
will reduce the volume of data reported and the i/o statistics will be averaged (interpolated) over
the reporting interval (unless the
.B \-u
option is specified, see below).
.TP
.B \-P
This indicates the number of decimals to print. The default precision \f2N\f1 may be set to something other than the
default 2 decimals.
.TP
.B \-u
When replaying an archive, by default values are reported
according to the selected sample interval (\c
.B \-t
option), not according to the actual record intervals in the archive.
To this effect PCP interpolates the values to be reported based on the
records in the archive, and is particularly useful when the
.B \-t
option is used to replay an archive with a longer sampling interval than
the underlying interval the archive was originally recorded with.
With the
.B \-u
option,
uninterpolated reporting is enabled - every value is reported
according to the native recording interval in the archive.
When the
.B \-u
option is specified, the
.B \-t
option makes no sense and is incompatible because the replay interval
is always the same as the recording interval in the archive.
In addition,
.B \-u
only makes sense when replaying an archive, see
.B \-a
above,
and so if
.B \-u
is specified then
.B \-a
must also be specified.
.TP
.B \-Z
By default,
.B pcp-iostat
reports the time of day according to the local timezone on the
system where
.B pcp-iostat
is run.
The
.B \-Z
option changes the timezone to
.I timezone
in the format of the environment variable
.B TZ
as described in
.BR environ (7).
.TP
.B \-z
Change the reporting timezone to the local timezone at the host that is
the source of the performance metrics, as identified via either the
.B \-h
or
.B \-a
options.
.TP
.B \-x
Specifies a comma separated list of one or more extended reporting options as follows:
.br
\f3dm\fP - report statistics for device-mapper logical devices instead of scsi devices,
.br
\f3t\fP - prefix every line in the report with a timestamp in \f2ctime\fP(3) format,
.br
\f3H\fP - omit the heading, which is otherwise reported every 24 samples.
.SH REPORT
The columns in the
.B pcp-iostat
report have the following interpretation :
.TP
.B Timestamp
When the \f3-x t\fP option is specified, this column is the timestamp in \f3ctime\fP(3) format.
.TP
.B Device
Specifies the scsi device name, or if \f3-x dm\fP is specified, the device-mapper logical device name.
.TP
.B rrqm/s
The number of read requests expressed as a rate per-second that were merged
during the reporting interval by the I/O scheduler.
.TP
.B wrqm/s
The number of write requests expressed as a rate per-second that were merged
during the reporting interval by the I/O scheduler.
.TP
.B r/s
The number of read requests completed by the device (after merges), expressed as a rate per second during the reporting interval.
.TP
.B w/s
The number of write requests completed by the device (after merges), expressed as a rate per second during the reporting interval.
.TP
.B rkB/s
The average volume of data read from the device expressed as KBytes/second during the reporting interval.
.TP
.B wkB/s
The average volume of data written to the device expressed as KBytes/second during the reporting interval.
.TP
.B avgrq-sz
The average I/O request size for both reads and writes to the device expressed as Kbytes during the reporting interval.
.TP
.B avgqu-sz
The average queue length of read and write requests to the device during the reporting interval.
.TP
.B await
The average time in milliseconds that read and write requests were queued (and serviced) to the device during the reporting interval.
.TP
.B r_await
The average time in milliseconds that read requests were queued (and serviced) to the device during the reporting interval.
.TP
.B w_await
The average time in milliseconds that write requests were queued (and serviced) to the device during the reporting interval.
.TP
.B %util
The percentage of time during the reporting interval that the device was busy processing requests.
A value of 100% indicates device saturation.
.SH FILES
.PD 0
.TP 10
.BI $PCP_VAR_DIR/pmns/ *
default PMNS specification files
.PD
.SH "PCP ENVIRONMENT"
Environment variables with the prefix
.B PCP_
are used to parameterize the file and directory names
used by PCP.
On each installation, the file
.I /etc/pcp.conf
contains the local values for these variables.
The
.B $PCP_CONF
variable may be used to specify an alternative
configuration file,
as described in
.BR pcp.conf (5).
.SH SEE ALSO
.BR pcp (1),
.BR PCPIntro (1),
.BR iostat2pcp (1),
.BR pmcd (1),
.BR pmchart (1),
.BR pmlogger (1),
.BR pcp.conf (5)
and
.BR pcp.env (5).
.SH DIAGNOSTICS
All are generated on standard error and are intended to be self-explanatory.


src/pcp/iostat/pcp-iostat.py
#!/usr/bin/pcp python
#
# Copyright (C) 2014-2015 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Iostat Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# pylint: disable=C0103,R0914,R0902
""" Display disk and device-mapper I/O statistics """

import sys
import signal
from pcp import pmapi, pmcc
from cpmapi import PM_TYPE_U64, PM_CONTEXT_ARCHIVE, PM_SPACE_KBYTE, PM_MODE_FORW

# use default SIGPIPE handler to avoid broken pipe exceptions
signal.signal(signal.SIGPIPE, signal.SIG_DFL)

IOSTAT_SD_METRICS = [ 'disk.dev.read', 'disk.dev.read_bytes',
                 'disk.dev.write', 'disk.dev.write_bytes',
                 'disk.dev.read_merge', 'disk.dev.write_merge',
                 'disk.dev.blkread', 'disk.dev.blkwrite',
                 'disk.dev.read_rawactive', 'disk.dev.write_rawactive',
                 'disk.dev.avactive']

IOSTAT_DM_METRICS = [ 'disk.dm.read', 'disk.dm.read_bytes',
                 'disk.dm.write', 'disk.dm.write_bytes',
                 'disk.dm.read_merge', 'disk.dm.write_merge',
                 'disk.dm.blkread', 'disk.dm.blkwrite',
                 'disk.dm.read_rawactive', 'disk.dm.write_rawactive',
                 'disk.dm.avactive']

class IostatReport(pmcc.MetricGroupPrinter):
    Hcount = 0
    def timeStampDelta(self, group):
        c = 1000000.0 * group.timestamp.tv_sec + group.timestamp.tv_usec
        p = 1000000.0 * group.prevTimestamp.tv_sec + group.prevTimestamp.tv_usec
        return (c - p) / 1000000.0

    def instlist(self, group, name):
        return dict(map(lambda x: (x[1], x[2]), group[name].netValues)).keys()

    def curVals(self, group, name):
        return dict(map(lambda x: (x[1], x[2]), group[name].netValues))

    def prevVals(self, group, name):
        return dict(map(lambda x: (x[1], x[2]), group[name].netPrevValues))

    def report(self, manager):
        precision = IostatOptions.Pflag
        if 'dm' in IostatOptions.xflag:
            subtree = 'disk.dm'
        else:
            subtree = 'disk.dev'
        group = manager["iostat"]

        if group[subtree + '.read_merge'].netPrevValues == None:
            # need two fetches to report rate converted counter metrics
            return

        instlist = self.instlist(group, subtree + '.read')
        dt = self.timeStampDelta(group)
        timestamp = group.contextCache.pmCtime(int(group.timestamp)).rstrip()

        c_rrqm = self.curVals(group, subtree + '.read_merge')
        p_rrqm = self.prevVals(group, subtree + '.read_merge')

        c_wrqm = self.curVals(group, subtree + '.write_merge')
        p_wrqm = self.prevVals(group, subtree + '.write_merge')

        c_r = self.curVals(group, subtree + '.read')
        p_r = self.prevVals(group, subtree + '.read')

        c_w = self.curVals(group, subtree + '.write')
        p_w = self.prevVals(group, subtree + '.write')

        c_rkb = self.curVals(group, subtree + '.read_bytes')
        p_rkb = self.prevVals(group, subtree + '.read_bytes')

        c_wkb = self.curVals(group, subtree + '.write_bytes')
        p_wkb = self.prevVals(group, subtree + '.write_bytes')

        c_ractive = self.curVals(group, subtree + '.read_rawactive')
        p_ractive = self.prevVals(group, subtree + '.read_rawactive')

        c_wactive = self.curVals(group, subtree + '.write_rawactive')
        p_wactive = self.prevVals(group, subtree + '.write_rawactive')

        c_avactive = self.curVals(group, subtree + '.avactive')
        p_avactive = self.prevVals(group, subtree + '.avactive')

        if "t" in IostatOptions.xflag:
            headfmt = "%-24s %-12s %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s"
            valfmt = "%-24s %-12s %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f"
        else:
            headfmt = "%-12s %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s"
            valfmt = "%-12s %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f %*.*f"

        if "h" not in IostatOptions.xflag:
            self.Hcount += 1
            if self.Hcount == 24:
                self.Hcount = 1
            if self.Hcount == 1:
                if "t" in IostatOptions.xflag:
                    heading = ('# Timestamp', 'Device',precision+5, 'rrqm/s',precision+5, 'wrqm/s',precision+5, 'r/s',precision+4,\
                    'w/s',precision+6, 'rkB/s',precision+6, 'wkB/s', precision+6,'avgrq-sz',precision+6, 'avgqu-sz',precision+5, \
                    'await',precision+5, 'r_await', precision+5,'w_await',precision+5, '%util')
                else:
                    heading = ('# Device',precision+5, 'rrqm/s',precision+5, 'wrqm/s',precision+5, 'r/s',precision+4, 'w/s'\
                    ,precision+6, 'rkB/s',precision+6, 'wkB/s', precision+6,'avgrq-sz',precision+6, 'avgqu-sz',precision+5,\
                    'await',precision+5, 'r_await',precision+5, 'w_await',precision+5, '%util')
                print(headfmt % heading)

        if p_rrqm == {} or p_wrqm == {} or p_r == {} or p_w == {} or \
           p_ractive == {} or p_wactive == {} or p_avactive == {} or \
           p_rkb == {} or p_wkb == {}:
            # no values for some metric (e.g. near start of archive)
            if "t" in IostatOptions.xflag:
                print(headfmt % (timestamp, 'NODATA', '?', '?', '?', '?', '?', '?','?', '?', '?', '?', '?', '?'))
            return

        try:
            for inst in sorted(instlist):
                # basic stats
                rrqm = (c_rrqm[inst] - p_rrqm[inst]) / dt
                wrqm = (c_wrqm[inst] - p_wrqm[inst]) / dt
                r = (c_r[inst] - p_r[inst]) / dt
                w = (c_w[inst] - p_w[inst]) / dt
                rkb = (c_rkb[inst] - p_rkb[inst]) / dt
                wkb = (c_wkb[inst] - p_wkb[inst]) / dt

                # totals
                tot_rios = (float)(c_r[inst] - p_r[inst])
                tot_wios = (float)(c_w[inst] - p_w[inst])
                tot_ios = (float)(tot_rios + tot_wios)

                # total active time in seconds (same units as dt)
                tot_active = (float)(c_avactive[inst] - p_avactive[inst]) / 1000.0

                avgrqsz = avgqsz = await = r_await = w_await = util = 0.0

                # average request size units are KB (sysstat reports in units of sectors)
                if tot_ios:
                    avgrqsz = (float)((c_rkb[inst] - p_rkb[inst]) + (c_wkb[inst] - p_wkb[inst])) / tot_ios

                # average queue length
                avgqsz = (float)((c_ractive[inst] - p_ractive[inst]) + (c_wactive[inst] - p_wactive[inst])) / dt / 1000.0

                # await, r_await, w_await
                if tot_ios:
                    await = ((c_ractive[inst] - p_ractive[inst]) + (c_wactive[inst] - p_wactive[inst])) / tot_ios

                if tot_rios:
                    r_await = (c_ractive[inst] - p_ractive[inst]) / tot_rios

                if tot_wios:
                    w_await = (c_wactive[inst] - p_wactive[inst]) / tot_wios

                # device utilization (percentage of active time / interval)
                if tot_active:
                    util = 100.0 * tot_active / dt

                device = inst	# prepare name for printing
                badcounters = rrqm < 0 or wrqm < 0 or r < 0 or w < 0 or await < 0 or avgrqsz < 0 or avgqsz < 0 or util < 0
                if precision == 1:
                    utilspace=precision+5
                    avgrqszspace=precision+7
                    awaitspace=precision+6
                else:
                    utilspace=precision+5
                    avgrqszspace=precision+6
                    awaitspace=precision+5
                if "t" in IostatOptions.xflag:
                    if badcounters:
                        print(headfmt % (timestamp, device, '?', '?', '?', '?', '?', '?', '?', '?', '?', '?', '?', '?'))
                    else:
                        print(valfmt % (timestamp, device,precision+5, precision, rrqm,precision+5,precision, wrqm,precision+4,precision,\
                        r,precision+4,precision, w,precision+6,precision, rkb,precision+6,precision, wkb, avgrqszspace,precision+1 ,avgrqsz,\
                        avgrqszspace,precision+1, avgqsz,precision+5,precision, await,awaitspace,precision, r_await,awaitspace,precision,\
                        w_await,utilspace,precision, util))
                else:
                    if badcounters:
                        print(headfmt % (device, '?', '?', '?', '?', '?', '?', '?', '?', '?', '?', '?', '?'))
                    else:
                        print(valfmt % (device,precision+5, precision, rrqm,precision+5,precision, wrqm,precision+5,precision, r,precision+4,\
                        precision, w,precision+6,precision, rkb,precision+6,precision, wkb,\
                        avgrqszspace,precision+1 ,avgrqsz,avgrqszspace,precision+1, avgqsz,precision+5,precision, await,awaitspace,precision,\
                        r_await,awaitspace,precision, w_await,utilspace,precision, util))
        except KeyError:
            # instance missing from previous sample
            pass

class IostatOptions(pmapi.pmOptions):
    # class attributes
    xflag = [] 
    uflag = None
    Pflag = 2

    def checkOptions(self, manager):
        if IostatOptions.uflag:
            if manager._options.pmGetOptionInterval():
                print("Error: -t incompatible with -u")
                return False
            if manager.type != PM_CONTEXT_ARCHIVE:
                print("Error: -u can only be specified with -a archive")
                return False
        return True

    def extraOptions(self, opt, optarg, index):
        if opt == "x":
            IostatOptions.xflag += optarg.replace(',', ' ').split(' ')
        elif opt == "u":
            IostatOptions.uflag = True
        elif opt == "P":
            IostatOptions.Pflag = int(optarg)

    def __init__(self):
        pmapi.pmOptions.__init__(self, "A:a:D:h:O:P:S:s:T:t:uVZ:z?x:")
        self.pmSetOptionCallback(self.extraOptions)
        self.pmSetLongOptionHeader("General options")
        self.pmSetLongOptionAlign()
        self.pmSetLongOptionArchive()
        self.pmSetLongOptionDebug()
        self.pmSetLongOptionHost()
        self.pmSetLongOptionOrigin()
        self.pmSetLongOption("precision", 1, "P", "N", "N digits after the decimal separator")
        self.pmSetLongOptionStart()
        self.pmSetLongOptionSamples()
        self.pmSetLongOptionFinish()
        self.pmSetLongOptionInterval()
        self.pmSetLongOption("no-interpolation", 0, "u", "", "disable interpolation mode with archives")
        self.pmSetLongOptionVersion()
        self.pmSetLongOptionTimeZone()
        self.pmSetLongOptionHostZone()
        self.pmSetLongOptionHelp()
        self.pmSetLongOptionHeader("Extended options")
        self.pmSetLongOption("", 1, 'x', "LIST", "comma separated extended options: [[dm],[t],[h]]")
        self.pmSetLongOptionText("\t\tdm\tshow device-mapper statistics (default is sd devices)")
        self.pmSetLongOptionText("\t\tt\tprecede every line with a timestamp in ctime format");
        self.pmSetLongOptionText("\t\th\tsuppress headings");

if __name__ == '__main__':
    try:
        opts = IostatOptions()
        manager = pmcc.MetricGroupManager.builder(opts, sys.argv)
        if not opts.checkOptions(manager):
            raise pmapi.pmUsageErr

        if IostatOptions.uflag:
            # -u turns off interpolation
            manager.pmSetMode(PM_MODE_FORW, manager._options.pmGetOptionOrigin(), 0)

        if "dm" in IostatOptions.xflag :
            manager["iostat"] = IOSTAT_DM_METRICS
        else:
            manager["iostat"] = IOSTAT_SD_METRICS
        manager.printer = IostatReport()
        sts = manager.run()
        sys.exit(sts)
    except pmapi.pmErr as error:
        sys.stderr.write('%s: %s\n' % (error.progname(), error.message()))
    except pmapi.pmUsageErr as usage:
        usage.message()
        sys.exit(1)
    except KeyboardInterrupt:
        pass
